blog/2024-05-20-standing-up-a-fhir-gateway-for-dhis2-from-an-ig/index.md

```diff
@@ -0,0 +1,553 @@
+        } catch (Exception e) {
+            throw new RuntimeCamelException(e);
+        }
+    }
+}
+```
+
+The code inside the constructor loads the IG within `package.tgz`. The `evaluate(...)` method, which is called from the Camel route, extracts the JSON payload from the message and executes the StructureMap https://dhis2.org/fhir/StructureMap/TrackedEntityToBundle to transform the tracked entity in the payload into a QuestionnaireResponse. Most of this code references classes in the [HL7 FHIR Core reference implementation](https://github.com/hapifhir/org.hl7.fhir.core). The reference implementation can be added to Java classpath by inserting the following dependency in the project's `pom.xml`:
+
+```xml
+<dependency>
+    <groupId>ca.uhn.hapi.fhir</groupId>
+    <artifactId>org.hl7.fhir.validation</artifactId>
+    <version>6.3.4</version>
+</dependency>
+```
+
+### Processing the FHIR QuestionnaireResponse Request
+
+We now wire the QuestionnaireResponse endpoint configured in `api.yaml` to a new Camel route which processes the FHIR request. Delete the `HelloWorldRouteBuilder.java` file in the package `org.hisp.dhis.integration.fhir.route` and create a new route builder class named `FhirGatewayRouteBuilder`, located within the same package. Add a route to the class as shown below:
+
+```java
+package org.hisp.dhis.integration.fhir.route;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.builder.RouteBuilder;
+import org.hisp.dhis.integration.fhir.FhirMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+
+@Component
+public class FhirGatewayRouteBuilder extends RouteBuilder {
+    
+  @Autowired
+  protected FhirMapper fhirMapper;
+
+  @Override
+  public void configure() throws Exception {
+      
+    from("direct:readQuestionnaireResponse")
+        .setHeader("CamelDhis2.queryParams", () -> Map.of("program", "IpHINAT79UW", "ouMode", "ACCESSIBLE"))
+        .toD("dhis2://get/resource?path=tracker/trackedEntities/${header.rid}&fields=*&client=#dhis2Client")
+        .transform(fhirMapper);
+    
+  }
+}
+```
+
+The route fetches the tracked entity from DHIS2, with the help of the DHIS2 component. `setHeader` is setting the query parameters to be included in the request sent out from the DHIS2 endpoint, that is, it sets the program ID query parameter to `IpHINAT79UW` and the organisation unit mode query parameter to `ACCESSIBLE`. The outbound DHIS2 endpoint configured in `toD("dhis2://get/resource?path=tracker/trackedEntities/${header.rid}&fields=*&client=#dhis2Client")` is fetching a tracked entity by ID. The ID is a variable referencing a Camel message header as denoted in `${header.rid}`. The header `rid` is a path parameter bound to the QuestionnaireResponse ID.
+
+:::tip
+Head over the [DHIS2 developer docs](https://developers.dhis2.org/docs/integration/camel-dhis2-component) for a primer into the Camel DHIS2 Component.
```

Comment:
Readers are already on the developers docs, you can just reference the doc page itself using `/docs/integration/camel-dhis2-component`. If you want to open it in a new tab instead, use html with `<a target="_blank" href="/docs/integration/camel-dhis2-component">` instead of using the full url to make sure the links works on elsewhere hosted instances too (such as localhost or netlify)